firmware/hw_layer/debounce.cpp

```diff
@@ -0,0 +1,46 @@
+/**
+ * @file        debounce.cpp
+ * @brief       Generic button debounce class
+ *
+ * @date Aug 31, 2020
+ * @author David Holdeman, (c) 2020
+ */
+#include "debounce.h"
+
+void ButtonDebounce::init (int t, brain_pin_e p, pin_input_mode_e m) {
+    threshold = MS2NT(t);
+    timeLast = 0;
+    pin = p;
+#ifdef PAL_MODE_INPUT_PULLDOWN
+    // getInputMode converts from pin_input_mode_e to iomode_t
+    mode = getInputMode(m);
+    efiSetPadMode("Button", p, mode);
+#endif
+}
+
+bool ButtonDebounce::readPinEvent() {
+    if (!pin) {
+        return false;
+    }
+    efitick_t timeNow = getTimeNowNt();
+    // If it's been less than the threshold since we were last called
+    if ((timeNow - timeLast) < threshold) {
+        return false;
```

Comment:
do we want to return false, or return the button's last state?

Reply:
for catching button events like this, we want to return false.

Reply:
Eventually, ButtonDebounce will need other methods that behave differently, but this was all I needed and all I'd be testing.

Reply:
ok, so it's not just a debouncer, but is also an edge detector, yes?